onnx/defs/tensor/defs.cc

```diff
@@ -371,6 +373,80 @@ ONNX_OPERATOR_SET_SCHEMA(
           ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
         }));
 
+static const char* Dim_ver15_doc = R"DOC( 
+Returns a specific dimension of the given input tensor's shape.  
+Takes an input tensor and the index (axis) of the dimension to extract (as an attribute) and
+returns the dimension value at the requested index (axis) in the input shape.
+The op returns the output as a 1-dimensional tensor of size 1,
+for the common usage scenario of creating other shapes (as 1-dimensional tensors).
+The axis specified must be a value in the range [-r, r-1] where r = rank(input).
+Otherwise, the output value is undefined.
+ 
+For example: 
+Input tensor with shape: [2, 3, 4] 
+Index: 1
+Output: [3] 
+ 
+Input tensor with shape: [2, 3, 4] 
+Index: 2
+Output: [4] 
+ 
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Dim,
+    15,
+    OpSchema()
+        .SetDoc(Dim_ver15_doc)
+        .Input(0, "input", "An input tensor.", "T", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
+        .Output(
+            0,
+            "output",
+            "Dimension of the input tensor at the specified axis.",
+            "tensor(int64)",
+            OpSchema::Single,
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .Attr(
+            "axis",
+            "The axis whose dimension is required. A negative value means counting axes from the back. "
+            "Accepted range is [-r, r-1] where r = rank(input).",
+            AttributeProto::INT)
+        .TypeConstraint("T", OpSchema::all_tensor_types_with_bfloat(), "Input tensor can be of arbitrary type.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          auto* type = ctx.getOutputType(0)->mutable_tensor_type();
+          type->set_elem_type(TensorProto::INT64);
+          // output of this operator is a 1D tensor of length 1
+          auto* shape = type->mutable_shape();
+          shape->add_dim()->set_dim_value(1);
+
+          if (hasInputShape(ctx, 0)) {
+            // Check that axis attribute has a proper value.
+            int rank = getInputShape(ctx, 0).dim_size();
+            int axis = static_cast<int>(getAttribute(ctx, "axis", 0));
```

Comment:
the spec does not mention a default value... should this be changed to 1. first verify attr has a value and then validate the value?

Reply:
There is no default-value, it is a required attribute. The default-checker would already have validated that the attribute is specified. I used ```getAttribute(...)``` as a short-cut (that does the checks anyway), but the default-value won't be used. I could replace this code by ```ctx.getAttribute("axis")->i()``` which is equivalent, if that seems preferable. I am inclined to avoid redundant checks here.

Reply:
if checker is run on the model then yes this should be caught... my point was current code makes you think default value is 0 which is not right... maybe add a comment 

My preference is to use getAttribute, validate not nullptr (fail if nullptr) and fetch value like we do for other required attrs